### PR TITLE
Implement instance-level announcements

### PR DESCRIPTION
**Pull request by lowjoel**
Depends on #75, #76.

I need UI help >.<

### CODE REVIEW

File: app/views/admin/announcements/_form.html.slim [hunk 1]
```diff
@@ -0,0 +1,6 @@
+= simple_form_for @announcement, resource: :admin_announcement do |f|
```
Comment (jsyeo) [inline app/views/admin/announcements/_form.html.slim]: Eh? The `announcement#new` action didn't initialize `@announcement` will this work?

Comment (wangqiang1208) [inline app/views/admin/announcements/_form.html.slim]: Cancancan gem did all this for us :P

Comment (jsyeo) [inline app/views/admin/announcements/_form.html.slim]: Eh? So magical? Okay.

File: spec/features/global_announcements_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+require 'rails_helper'
+
+RSpec.describe 'Global announcements', type: :feature do
+  subject { page }
+  let(:instance) { create(:instance) }
+
+  with_tenant(:instance) do
+    describe 'no global announcements' do
+      before do
+        instance.announcements.clear
```
Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: I think you should extract this line out.

BTW, I still think we should clear db after factor girl lint, Then you'll not need to do this

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: It is better for tests requiring state to create that state, than make every test depend on state.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: Why not just make default state an empty one

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: That allows your tests to depend on state.

By default, the clear function call here does nothing. The instance has no announcements. I am just asserting a precondition for the test.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: But can we asserting this precondition globally ?  Then we no need to do it in each tests.

And even you want do it in tests, I think you should put this line in a up-layer before, because you are doing the same thing under every describe.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: By doing that you cannot run your tests in parallel.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: By doing this there are lots of duplication code in your tests.

And that's the meaning of nesting,  I don't think we can gain much from
running tests in parallel.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: It'd about correctness of your tests. By clearing all your state you require that all the tests be run. This removes the isolation between tests. What's worse is you get tests that pass and fail randomly.

See http://blog.stevensanderson.com/2009/08/24/writing-great-unit-tests-best-and-worst-practises/

Mock out all external services and state
Otherwise, behaviour in those external services overlaps multiple tests, and state data means that different unit tests can influence each other’s outcome.
You’ve definitely taken a wrong turn if you have to run your tests in a specific order, or if they only work when your database or network connection is active.
(By the way, sometimes your architecture might mean your code touches static variables during unit tests. Avoid this if you can, but if you can’t, at least make sure each test resets the relevant statics to a known state before it runs.)

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: Rspec did pretty much of this for you, by nesting your tests are still isolated, there is no depending on each other and the order is random, it's actually doing the same thing as what you are doing there.

In this case it still call <code>instance.annoucements.clear</code> for 3 times, and things under describe runs in a random order.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: You're asking about clearing the database. It's two different things were talking about here.
If it can be factored out, sure. But clearing the database at the start of all tests is not the solution for writing state preconditions.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: ....well, I'm also confused, I'm talking about factored that line out.

Now for clearing database:
>By clearing all your state you require that all the tests be run. This removes the isolation between tests. 
I don't understand here, why clearing the database removes the isolation ?

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: It does not remove isolation. But by clearing the database it is easier to write tests that depend on state.

Furthermore, clearing the database takes time, at least 5 seconds per run. If you're running tests regularly, like you should, it adds up. Running tests in parallel also helps when you have hundreds of tests in future.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: And I've looked at the code again, clearing instance announcements is not repetitive. It is setting up the preconditions needed for the test. It looks like it's the same but it's testing different things.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: I mean you factor it out, it  still do the same thing for you. And after that you no need to build announcements and save, can just create.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: they are separate tests. they are asserting their own preconditions. factoring out the clear statement would bring the required state outside of the local scope and will be missed when more specs are added in future, or when that precondition is not required for the series of tests. yes, the effect is the same, but the confusion here is for the programmer not the program.

create also cannot be used, because the records must be created before visiting the page. this is one of the problems of using before statements.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: Don't think there's any confusion, it's pretty convention, when tests added which does not require that precondition you should do :
```ruby
describe 'precondition' do
  before { some precondition }
  context 'xxx' do
  end

  context 'xx' do
  end
end

describe 'other preconditions or no precondition' do
  context 'xxx' 
end
```

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: I considered that. Put your suggestion into practice on this spec and you'll see that the descriptions are weird.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: And for create
```ruby
let!(:some_var) { create(:something) }
```
can create things for you before you visiting the page

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: Yes. All these are best practices. Best practices that, IMO, are being applied without considering the circumstances. Engineering is not only about following rules. It's knowing when exceptions need to be made when other considerations are in play.

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: >I considered that. Put your suggestion into practice on this spec and you'll see that the descriptions are weird.

I think you can refine your descriptions or just put things depends on certain precondition into one spec file.

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: Suggest replacements please?

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: >Engineering is not only about following rules. It's knowing when exceptions need to be made when other considerations are in play.

We should not always follow rules that's true, but I don't think this is the case we should break rules..

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: >Suggest replacements please?
http://www.codeshare.io/XSJo1
In this case since all 3 describe have same precondition, can factored out and don't need a extra describe.

When we have tests which don't depend on this precondition, I think we should create a new file for them. ( and in this case we will probably not have those tests right ? )

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: The suggested solution does not address my concern that all the tests now depend on an empty set of instance announcements. Adding another test now will have that test start with an empty set of announcements, which is stateful even if the test does not require that precondition.

Your solution to that is to make a new file. I think that is also a bad idea because now the specs for one feature is split over two places.

You might not think that this is important, but I do. I've seen projects grow and I've seen code bases evolve. There's always unquantifiable changes that happen to code bases over time and I'm trying to make sure as few ways these things can happen. If you think that your suggestion is still better in the long run, I'll leave the discussion right here and adopt your solution.

Comment (weiqingtoh) [inline spec/features/global_announcements_spec.rb]: @lowjoel @wangqiang1208 , 

I've been following the conversation on both ends. I also agree with both sides - on Joel's end it is about correctness, right assertion, and code readability. On Wang Qiang's end, it is about abstracting repetitive code as much as possible. The hard part is how to choose which - especially in this situation when it gets a little grey. 

I learn towards the argument that readability is compromised if we abstract certain pre-conditions. To the experienced developer, either style should be fine. However, to the new RoR guy, reading tests with abstracted pre-conditions can be quite tough.  

If we consider each test to be an independent unit, then the specs should have the full, relevant set of clearly stated pre-conditions. *In this case, the relevant pre-conditions are mainly the models/controllers you are trying to test - usually announcements, and not the user authentication or account creation.*  

Wang Qiang, if you still not convinced maybe you can suggest a different way of writing the entire file and we can discuss?

----

Just for kicks, I ran `rspec spec/features/global_announcements_spec.rb` twice for both cases - results are below. Not the best method, but this was a quick and dirty way to measure the time taken.

##### Joel's suggestion (Slightly longer average of 0.04 seconds)
- `Finished in 0.62794 seconds (files took 1.86 seconds to load)`
- `Finished in 0.6446 seconds (files took 1.89 seconds to load)`

##### Wang Qiang's suggestion
- `Finished in 0.63257 seconds (files took 1.82 seconds to load)`
- `Finished in 0.63079 seconds (files took 1.84 seconds to load)`

Comment (wangqiang1208) [inline spec/features/global_announcements_spec.rb]: > I've seen projects grow and I've seen code bases evolve. There's always unquantifiable changes that happen to code bases over time and I'm trying to make sure as few ways these things can happen.
@lowjoel I think next time we should sit down together to further discuss this and I think I'll learn more from you, I'm still not fully aware about the things you said here :(   So for now you don't need to adopt my solution.

@weiqingtoh Thanks for the detailed comments ! >< Actually I think abstraction might not have impact on code readability cause it's the Rspec's conventional way and it improves maintainability. Well, same, next time we can discuss further offline, let's end the discussion here.

**Comment by lowjoel** [comment]
@jsyeo I should be rebasing this... right?

**Comment by jsyeo** [comment]
Yes please.

**Comment by lowjoel** [comment]
@jsyeo omataseshimashita

File: app/controllers/admin/announcements_controller.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+class Admin::AnnouncementsController < Admin::Controller
```
Comment (lowjoel) [inline app/controllers/admin/announcements_controller.rb]: @jsyeo I'm quite certain this is a false positive?

File: app/views/admin/announcements/_announcement.html.slim [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  h2
+    - if announcement.unread?
+      input.new_announcement type="hidden"
+      = fa_icon 'envelope'
+    = announcement.title
+  - if can? :manage, announcement.class
+    div.pull-right
+      = link_to edit_admin_announcement_path(announcement), class: 'btn btn-default' do
+        = fa_icon 'edit'
+      end
```
Comment (wangqiang1208) [inline app/views/admin/announcements/_announcement.html.slim]: <code>end</code> again :-P

File: app/views/admin/announcements/index.html.slim [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  h1
+    | Instance Announcements
+    - if can?(:manage, Instance::Announcement)
+      div.pull-right
+        = link_to(t('.new'),
+          new_admin_announcement_path,
+          class: ['btn', 'btn-primary'])
+- @announcements.each do |announcement|
+  = render partial: 'announcement', locals: { announcement: announcement }
+end
```
Comment (wangqiang1208) [inline app/views/admin/announcements/index.html.slim]: same here

**Comment by lowjoel** [comment]
oops! nice catch.

File: app/views/admin/announcements/index.html.slim [hunk 6]
```diff
@@ -0,0 +1,10 @@
+div.page-header
+  h1
+    | Instance Announcements
+    - if can?(:manage, Instance::Announcement)
+      div.pull-right
+        = link_to(t('.new'),
+          new_admin_announcement_path,
+          class: ['btn', 'btn-primary'])
+- @announcements.each do |announcement|
```
Comment (wangqiang1208) [inline app/views/admin/announcements/index.html.slim]: I think <code>render @announcements</code> will do ? Or you prefer this for some reason ?

File: app/views/admin/announcements/_announcement.html.slim [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    = announcement.title
+  - if can? :manage, announcement.class
+    div.pull-right
+      = link_to edit_admin_announcement_path(announcement), class: 'btn btn-default' do
+        = fa_icon 'edit'
+      = link_to admin_announcement_path(announcement), class: ['btn', 'btn-danger'],
+        method: 'DELETE', data: { confirm: 'lol' } do
+        = fa_icon 'trash'
+  h3.timestamp
+    => announcement.valid_from.to_formatted_s(:long)
```
Comment (wangqiang1208) [inline app/views/admin/announcements/_announcement.html.slim]: We already changed to use a helper here <code>format_datetime</code>

File: app/views/admin/announcements/_announcement.html.slim [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    - if announcement.unread?
+      input.new_announcement type="hidden"
+      = fa_icon 'envelope'
+    = announcement.title
+  - if can? :manage, announcement.class
+    div.pull-right
+      = link_to edit_admin_announcement_path(announcement), class: 'btn btn-default' do
+        = fa_icon 'edit'
+      = link_to admin_announcement_path(announcement), class: ['btn', 'btn-danger'],
+        method: 'DELETE', data: { confirm: 'lol' } do
```
Comment (wangqiang1208) [inline app/views/admin/announcements/_announcement.html.slim]: The confirm message, lol

File: config/locales/en.yml [hunk 9]
```diff
@@ -30,6 +30,30 @@ en:
...
+      more_announcements: More Announcements...
+  admin:
+    navbar:
+      announcements: Announcements
+    admin:
+      home:
+        message: "Pages"
+    announcements:
+      index:
+        header: "Announcements"
```
Comment (wangqiang1208) [inline config/locales/en.yml]: Do we restrict to use single or double quotes in this file ?

Comment (jsyeo) [inline config/locales/en.yml]: Since it's yaml, why not just drop the quotes?

File: app/views/admin/announcements/index.html.slim (hunk 6 above)
Comment (lowjoel) [inline app/views/admin/announcements/index.html.slim]: ah, I didn't know that. let's try.

File: config/locales/en.yml (hunk 9 above)
Comment (lowjoel) [inline config/locales/en.yml]: yaml evokes json... idk, I tend to use quotes. what should we stick with?

Comment (jsyeo) [inline config/locales/en.yml]: Hmmm... http://stackoverflow.com/a/22235064 says

> Use quotes if your value includes special characters, (e.g. :, {, }, [, ], ,, &, *, #, ?, |, -, <, >, =, !, %, @, \).
> "Yes" and "No" should be enclosed in quotes (single or double) or else they will be interpreted as TrueClass and FalseClass values:

Since I expect `#{...}` to be very common, let's stick to using quotes. How about single quotes? Since we are already preferring single quotes in ruby.

Comment (lowjoel) [inline config/locales/en.yml]: Single quotes :+1:

also saves pressing the shift key. (sorry :smile:)

Comment (lowjoel) [inline config/locales/en.yml]: Maybe we should just accept this first, merge it, then replace all to single quotes in one fell swoop. the other files are missing convention also.

Comment (wangqiang1208) [inline config/locales/en.yml]: OK, I'll update them after other PRs merged.
